File: 프로그래머스/1/172928. 공원 산책/공원 산책.py
def solution(park, routes):
    x, y = 0, 0
    for row in range(len(park)):
        for col in range(len(park[row])):
            if park[row][col] == 'S':  
                x, y = row, col
    
    op = {'N':(-1, 0), 'S':(1, 0), 'W':(0, -1), 'E':(0, 1)}
    
    for i in routes:
        dx, dy = op[i.split()[0]] 
        n = int(i.split()[1])  
        
        xx, yy = x, y 
        canmove = True 

        for _ in range(n):
            nx = xx + dx  
            ny = yy + dy 
            

            if 0 <= nx <= len(park)-1 and 0 <= ny <= len(park[0])-1 and park[nx][ny] != 'X':
                canmove = True
                xx, yy = nx, ny
            else:  # 공원을 벗어낫거나, 장애물이면 이동 불가(False)
                canmove = False
                break
                
        if canmove:  # 이동이 가능하면 위치 반영해주기
            x, y = nx, ny  
        
    return [x, y]

File: 프로그래머스/1/172928. 공원 산책/공원 산책.js
function solution(park, routes) {
    var answer = [];
    let currY = 0
    let currX = 0
    const pr = park.length;
    const pc = park[0].length;
    
    // 시작점찾기
    for(let i=0;i<pr;i++){
        for(let j=0;j<pc;j++){
            if(park[i][j]==="S") {
                currY = i
                currX = j
            }
        }
    }
    
    for(let routeIdx = 0; routeIdx<routes.length; routeIdx++){
        const route = routes[routeIdx].split(" ")
        const op = route[0] // 이동할 방향
        const n = Number(route[1]) // 이동할 칸 수 
        
        let ny = currY 
        let nx = currX
        

        let flag = false 
        for(let i = 0; i < n ;i++){
            if(op==="E") nx++;
            if(op==="S") ny++;
            if(op==="W") nx--;
            if(op==="N") ny--;
            
             // 공원 벗어나거나 장애물 만나면 무시
            if(ny < 0 || ny >= pr || nx < 0 || nx >= pc || park[ny][nx]==="X" ){
                flag = true 
                break;
            }
        }
        
        if(flag) continue
      
        
        //이동
        currY = ny
        currX = nx
    }
        
    return [currY, currX];
}